Add Integration Tests for Kopia RepositoryServer based functions (#2026)
kanisterio/kanister: examples/mysql-kopia-repository-server/README.md

# MySQL

[MySQL](https://MySQL.org) is one of the most popular database servers in the world. Notable users include Wikipedia, Facebook and Google.

## Introduction

This chart bootstraps a single node MySQL deployment on a [Kubernetes](http://kubernetes.io) cluster using the [Helm](https://helm.sh) package manager.

## Prerequisites

- Kubernetes 1.20+
- PV provisioner support in the underlying infrastructure
- Kanister controller version 0.90.0 installed in your cluster, let's assume in Namespace `kanister`
- Kanctl CLI installed (https://docs.kanister.io/tooling.html#install-the-tools)

## Installing the Chart

To install the MySQL database using the `bitnami` chart with the release name `mysql-release`:

```bash
# Add bitnami in your local chart repository
$ helm repo add bitnami https://charts.bitnami.com/bitnami

# Update your local chart repository
$ helm repo update

# Install the MySQL database

> 

Reviewer: let's add a note here, specifying why are we using custom image here.

Author: Added !

$ helm install mysql-release bitnami/mysql \
  --namespace mysql \
  --create-namespace \
  --set image.repository=kanisterio/mysql \
  --set image.tag=0.90.1 \
  --set auth.rootPassword='<mysql-root-password>'
```

The command deploys a MySQL instance in the `mysql` namespace.

By default, a random password will be generated for the root user. For setting your own password, use the `auth.rootPassword` param as shown above.

You can retrieve your root password by running the following command. Make sure to replace [YOUR_RELEASE_NAME] and [YOUR_NAMESPACE]:

    `kubectl get secret [YOUR_RELEASE_NAME] --namespace [YOUR_NAMESPACE] -o jsonpath="{.data.mysql-root-password}" | base64 --decode`

> **Tip**: List all releases using `helm list --all-namespaces`, using Helm Version 3.

## Integrating with Kanister

If you have deployed MySQL application with name other than `mysql-release` and namespace other than `mysql`, you need to modify the commands(backup, restore and delete) used below to use the correct release name and namespace

> ### Create Repository Server CR

Reviewer: ```suggestion
### Create Repository Server CR

```

Author: Done !

   type: czM=
   bucket: <base64_encoded_s3_bucket_name>
   # optional: used as a sub path in the bucket for all backups
   path: <base_64_encoded_prefix_provided_when_creating_kopia_repository>
   # required, if supported by the provider
   region: <base64_encoded_s3_bucket_region>
   # optional: if set to true, do not verify SSL cert.
   # Default, when omitted, is false
   #skipSSLVerify: false
   # required: if type is `file-store`
   # optional, otherwise
   #claimName: store-pvc
```

- Apply Secrets
```bash
$ kubectl create secret tls repository-server-tls-cert --cert=certificate.pem --key=key.pem -n kanister

kubectl create secret generic repository-server-user-access -n kanister --from-literal=localhost=<suitable_password_for_repository_server_user>

kubectl create secret generic repository-admin-user -n kanister --from-literal=username=admin@mysql --from-literal=password=<suitable_password_for_repository_server_admin>

kubectl create secret generic repo-pass -n kanister --from-literal=repo-password=<repository_password_set_while_creating_kopia_repository>

kubectl apply -f s3_location_creds.yaml -n kanister

kubectl apply -f s3_location.yaml -n kanister
```

> - Apply Repository Server CRD

Author: Done !

> - Apply Repository Server CRD

Reviewer: We need to change this, because the new CRD should automatically be installed with Kanister installation.

# MySQL

[MySQL](https://MySQL.org) is one of the most popular database servers in the world. Notable users include Wikipedia, Facebook and Google.

## Introduction

This chart bootstraps a single node MySQL deployment on a [Kubernetes](http://kubernetes.io) cluster using the [Helm](https://helm.sh) package manager.

## Prerequisites

- Kubernetes 1.20+
- PV provisioner support in the underlying infrastructure
- Kanister controller version 0.90.0 installed in your cluster, let's assume in Namespace `kanister`
- Kanctl CLI installed (https://docs.kanister.io/tooling.html#install-the-tools)

## Installing the Chart

To install the MySQL database using the `bitnami` chart with the release name `mysql-release`:

```bash
# Add bitnami in your local chart repository
$ helm repo add bitnami https://charts.bitnami.com/bitnami

# Update your local chart repository
$ helm repo update

# Install the MySQL database

> > **Note**: The custom image for MySQL is required in order to utilize the Kopia Repository Server based functions.

Reviewer: ```suggestion
> **Note**: The custom image for MySQL is required because the Kopia Repository Server based functions (used in Blueprint), require `Kopia` binary to be available on application pod.
```

$ helm install mysql-release bitnami/mysql \
  --namespace mysql \
  --create-namespace \
  --set image.repository=kanisterio/mysql-kopia \
  --set image.tag=0.90.1 \
  --set auth.rootPassword='<mysql-root-password>'
```

The command deploys a MySQL instance in the `mysql` namespace.

By default, a random password will be generated for the root user. For setting your own password, use the `auth.rootPassword` param as shown above.

You can retrieve your root password by running the following command. Make sure to replace [YOUR_RELEASE_NAME] and [YOUR_NAMESPACE]:

    `kubectl get secret [YOUR_RELEASE_NAME] --namespace [YOUR_NAMESPACE] -o jsonpath="{.data.mysql-root-password}" | base64 --decode`

> **Tip**: List all releases using `helm list --all-namespaces`, using Helm Version 3.

## Integrating with Kanister

If you have deployed MySQL application with name other than `mysql-release` and namespace other than `mysql`, you need to modify the commands(backup, restore and delete) used below to use the correct release name and namespace

### Create Repository Server CR

> Create Kopia Repository Server CR if not created already

Reviewer: Let's point them to our Kanister documentation that explains what this CR is and how this workflow looks like. Basically point them to the kanister docs that talks about Kopia Repo Server workflow in detail.

Author: I think we haven't added the docs yet for Kopia Repository Server CR

Author: Sure !

- Create Kopia Repository
```bash
$ kopia --log-level=error --config-file=/tmp/kopia-repository.config \
--log-dir=/tmp/kopia-cache repository create --no-check-for-updates \
--cache-directory=/tmp/cache.dir --content-cache-size-mb=0 \
--metadata-cache-size-mb=500 --override-hostname=mysql.app \
--override-username=kanisterAdmin s3 --bucket=<s3_bucket_name> \
--prefix=/repo-controller/ --region=<s3_bucket_region> \
--access-key=<aws_access_key> --secret-access-key=<aws_secret_access_key> --password=<repository_password>
```

- Create Certificates and Location Secrets for Kopia Repository Server
```bash
$ openssl req -newkey rsa:2048 -nodes -keyout key.pem -x509 -days 365 -out certificate.pem

> ```

Reviewer: It looks like we generated the cert here and then are creating the secret later. This is not very straight forward. Let's create each secret separately. 
For example, generate cert and immediately after that create secret. 
Then talk about the other secret yaml and then create that secret.

Author: Done !

    app.kubernetes.io/created-by: kanister
  name: kopia-repo-server-1
  namespace: kanister
spec:
  storage:
    secretRef:
      name: s3-location
      namespace: kanister
    credentialSecretRef:
      name: s3-creds
      namespace: kanister
  repository:
    rootPath: /repo-controller/
    passwordSecretRef:
      name: repo-pass
      namespace: kanister
    username: kanisterAdmin
    hostname: mysql.app
  server:
    adminSecretRef:
      name: repository-admin-user
      namespace: kanister
    tlsSecretRef:
      name: repository-server-tls-cert
      namespace: kanister
    userAccess:
      userAccessSecretRef:
        name: repository-server-user-access
        namespace: kanister
      username: kanisteruser
```
```bash
$ kubectl create -f repo-server-cr.yaml -n kanister
```
**NOTE:**

Make Sure the Repository Server is in ServerReady State before creating actionsets.
You could check the status of the Repository Server CR by running following command
```bash
$ kubectl get repositoryservers.cr.kanister.io kopia-repo-server-1 -n kanister -o yaml
```

**NOTE:**

The above command will configure a kopia repository server, which manages artifacts resulting from Kanister

> data operations such as backup should go. This is stored as a `repositoryservers.cr.kanister.io`

Reviewer: > The above command will configure a kopia repository server, which manages artifacts resulting from Kanister
> data operations such as backup should go.

This sentence is incorrect and you please fix this.

Author: Done !

*CustomResource (CR)* which is then referenced in Kanister ActionSets. Every ActionSet
requires a Repository Server reference to complete the action. This CR (`repositoryservers.cr.kanister.io`)

> can be shared between Kanister-enabled application instances.

Reviewer: > This CR (`repositoryservers.cr.kanister.io`)
> can be shared between Kanister-enabled application instances.

What do you mean by this?

Author: It means like we can use repository server with all the applications which are supported by Kanister, But I get your point.

### Create Blueprint

Create Blueprint in the same namespace as the Kanister controller

```bash
$ kubectl create -f ./mysql-blueprint.yaml -n kanister
```

Once MySQL is running, you can populate it with some data. Let's add a table called "pets" to a test database:

```bash
# Connect to MySQL by running a shell inside MySQL's pod
$ kubectl exec -ti $(kubectl get pods -n mysql --selector=app.kubernetes.io/instance=mysql-release -o=jsonpath='{.items[0].metadata.name}') -n mysql -- bash

# From inside the shell, use the mysql CLI to insert some data into the test database

> # Create "test" db

Reviewer: Is this a command to create the database? If yes, it's incorrect I think.
Should this not be after we have logged into the MySQL instance using mysql  command?

Author: This is just headline, the commands are below that

+----------+-------+---------+------+------------+-------+
| Puffball | Diane | hamster | f    | 1999-03-30 | NULL  |
+----------+-------+---------+------+------------+-------+
1 row in set (0.00 sec)
```

## Protect the Application

You can now take a backup of the MySQL data using an ActionSet defining backup for this application. Create an ActionSet in the same namespace as the controller.

```bash
# Find Repository name
$ kubectl get repositoryservers -n kanister
NAME                  AGE
kopia-repo-server-1   2m

# Create Actionset
# Please make sure the value of repository-server and blueprint matches with the names of repository-server and blueprint that we have created already
kanctl create actionset --action backup --namespace kanister --blueprint mysql-blueprint --statefulset mysql/mysql-release --secrets mysql=mysql/mysql-release --repository-server=kanister/kopia-repo-server-1
actionset backup-rslmb created

# View the status of the actionset
$ kubectl --namespace kanister get actionsets.cr.kanister.io backup-rslmb
NAME           PROGRESS   LAST TRANSITION TIME   STATE
backup-rslmb   100.00     2022-12-15T09:56:49Z   complete
```

### Disaster strikes!

> Let's say someone accidentally deleted the test database using the following command:
> ```bash

Reviewer: ```suggestion
Let's say someone accidentally deleted the test database using the following command:

```bash
```
Can you please do this at other places as well, a new line before the code block.

$ helm install mysql-release bitnami/mysql \
  --namespace mysql \
  --create-namespace \
  --set image.repository=kanisterio/mysql-kopia \
  --set image.tag=0.90.2 \
  --set auth.rootPassword='<mysql-root-password>'
```

The command deploys a MySQL instance in the `mysql` namespace.

By default, a random password will be generated for the root user. For setting your own password, use the `auth.rootPassword` param as shown above.

You can retrieve your root password by running the following command. Make sure to replace [YOUR_RELEASE_NAME] and [YOUR_NAMESPACE]:

    `kubectl get secret [YOUR_RELEASE_NAME] --namespace [YOUR_NAMESPACE] -o jsonpath="{.data.mysql-root-password}" | base64 --decode`

> **Tip**: List all releases using `helm list --all-namespaces`, using Helm Version 3.

## Integrating with Kanister

If you have deployed MySQL application with name other than `mysql-release` and namespace other than `mysql`, you need to modify the commands(backup, restore and delete) used below to use the correct release name and namespace

### Create Repository Server CR

Create Kopia Repository Server CR if not created already

> - Create Kopia Repository

Reviewer 2: ```suggestion
- Create Kopia Repository using S3 as the location storage
```

```bash
$ kopia --log-level=error --config-file=/tmp/kopia-repository.config \
--log-dir=/tmp/kopia-cache repository create --no-check-for-updates \
--cache-directory=/tmp/cache.dir --content-cache-size-mb=0 \
--metadata-cache-size-mb=500 --override-hostname=mysql.app \
--override-username=kanisterAdmin s3 --bucket=<s3_bucket_name> \
--prefix=/repo-controller/ --region=<s3_bucket_region> \
--access-key=<aws_access_key> --secret-access-key=<aws_secret_access_key> --password=<repository_password>
```

- Generate TLS Certificates and create TLS secret for Kopia Repository Server for secure communication between Kopia Repository Server and Client

```bash
$ openssl req -newkey rsa:2048 -nodes -keyout key.pem -x509 -days 365 -out certificate.pem

$ kubectl create secret tls repository-server-tls-cert --cert=certificate.pem --key=key.pem -n kanister
```

- Create Location Secrets for Kopia Repository

```bash
# The following file s3_location_creds.yaml is a sample file for creating s3 credentials secrets. It contains the credentials for accessing the s3 bucket.

> $ vi s3_location_creds.yaml

Reviewer 2: Can we specify at the start of the readme that we will be using S3 as the location storage

Author: Done !

apiVersion: v1
kind: Secret
metadata:
   name: s3-creds
   namespace: kanister
   labels:
      repo.kanister.io/target-namespace: monitoring
type: secrets.kanister.io/aws
data:
   # required: base64 encoded value for key with proper permissions for the bucket
   aws_access_key_id: <base64_encoded_aws_access_key>
   # required: base64 encoded value for the secret corresponding to the key above
   aws_secret_access_key: <base64_encoded_aws_secret_access_key>
```

```
$ kubectl create -f s3_location_creds.yaml -n kanister
```

```bash

> # The following file s3_location.yaml is a sample file for creating s3 location secrets. It contains the details of the s3 bucket

Reviewer: ```suggestion
# The following file s3_location.yaml is a sample file for creating s3 location secrets. It contains the details of the s3 bucket.
```

$ vi s3_location.yaml

apiVersion: v1
kind: Secret
metadata:
   name: s3-location
   namespace: kanister
   labels:
      repo.kanister.io/target-namespace: monitoring
type: Opaque
data:
   # required: specify the type of the store
   # supported values are s3, gcs, azure, and file-store

>    type: czM=

Reviewer 2: why are we specifying that this is base 64 encoded as we are specifying for other keys in this secret?

Author: its base64 encoded value of s3

Reviewer 2: My bad, we are using s3, hence the base64 encoded value. can you add a comment over there specifying its base64 encoded value of s3

   bucket: <base64_encoded_s3_bucket_name>
   # optional: used as a sub path in the bucket for all backups
   path: <base_64_encoded_prefix_provided_when_creating_kopia_repository>
   # required, if supported by the provider
   region: <base64_encoded_s3_bucket_region>
   # optional: if set to true, do not verify SSL cert.
   # Default, when omitted, is false
   #skipSSLVerify: false
   # required: if type is `file-store`
   # optional, otherwise
   #claimName: store-pvc
```

```
$ kubectl create -f s3_location.yaml -n kanister
```

- Apply Secrets for Kopia Repository Server User Access, Admin Access and Repository Access

```bash

> # The following command creates secrets for kopia repository server user access

Author: Done !

kubectl create secret generic repository-server-user-access -n kanister --from-literal=localhost=<suitable_password_for_repository_server_user>

> # The following command creates secrets for kopia repository server admin access

Author: Done !

> kubectl create secret generic repository-admin-user -n kanister --from-literal=username=admin@mysql --from-literal=password=<suitable_password_for_repository_server_admin>

Reviewer 2: ```suggestion
kubectl create secret generic repository-admin-user -n kanister --from-literal=username=<suitable_username_for_repository_server_admin> --from-literal=password=<suitable_password_for_repository_server_admin>
```

> # The following command creates secrets for kopia repository access

Author: Done !

kubectl create secret generic repo-pass -n kanister --from-literal=repo-password=<repository_password_set_while_creating_kopia_repository>
```

> - Create Repository Server CR

Author: Done !

    `kubectl get secret [YOUR_RELEASE_NAME] --namespace [YOUR_NAMESPACE] -o jsonpath="{.data.mysql-root-password}" | base64 --decode`

> **Tip**: List all releases using `helm list --all-namespaces`, using Helm Version 3.

## Integrating with Kanister

If you have deployed MySQL application with name other than `mysql-release` and namespace other than `mysql`, you need to modify the commands(backup, restore and delete) used below to use the correct release name and namespace

### Create Repository Server CR

Create Kopia Repository Server CR if not created already

- Create Kopia Repository
```bash
$ kopia --log-level=error --config-file=/tmp/kopia-repository.config \
--log-dir=/tmp/kopia-cache repository create --no-check-for-updates \
--cache-directory=/tmp/cache.dir --content-cache-size-mb=0 \
--metadata-cache-size-mb=500 --override-hostname=mysql.app \
--override-username=kanisterAdmin s3 --bucket=<s3_bucket_name> \
--prefix=/repo-controller/ --region=<s3_bucket_region> \
--access-key=<aws_access_key> --secret-access-key=<aws_secret_access_key> --password=<repository_password>
```

> - Generate TLS Certificates and create TLS secret for Kopia Repository Server for secure communication between Kopia Repository Server and Client

Reviewer: ```suggestion
- Generate TLS Certificates and create TLS secret for Kopia Repository Server for secure communication between Kopia Repository Server and Client

```
